backend/de.metas.payment.sepa/base/src/main/java/de/metas/payment/sepa/sepamarshaller/impl/SEPAVendorCreditTransferMarshaler_Pain_001_001_03_CH_02.java

```diff
@@ -684,23 +687,47 @@ else if (!Check.isEmpty(line.getSwiftCode(), true))
 				Check.errorIf(paymentType == PAYMENT_TYPE_1, SepaMarshallerException.class,
 						"SEPA_ExportLine {} has to have StructuredRemittanceInfo", createInfo(line));
 
-				// note: we use the structuredRemittanceInfo in ustrd, if we do SEPA (zahlart 5),
-				// because it's much less complicated
-				final String reference = Check.isEmpty(line.getStructuredRemittanceInfo(), true)
-						? line.getDescription()
-						: line.getStructuredRemittanceInfo();
 
-				// provide the line-description (if set) as unstructured remittance info
-				if (Check.isEmpty(reference, true))
+				if (!Check.isEmpty(bankAccount.getQR_IBAN(),true))
 				{
-					// at least add a "." to make sure the node exists.
-					rmtInf.setUstrd(".");
+					final String QRReference = StringUtils.cleanWhitespace(line.getStructuredRemittanceInfo());
+
+					Check.errorIf(isInvalidQRReference(QRReference), SepaMarshallerException.class,
+								  "SEPA_ExportLine {} has to valid QR Reference", createInfo(line));
+
+					final StructuredRemittanceInformation7 strd = objectFactory.createStructuredRemittanceInformation7();
+					rmtInf.setStrd(strd);
+					final CreditorReferenceInformation2 cdtrRefInf = objectFactory.createCreditorReferenceInformation2();
+					strd.setCdtrRefInf(cdtrRefInf);
+					final CreditorReferenceType2 tp = objectFactory.createCreditorReferenceType2();
+					cdtrRefInf.setTp(tp);
+					final CreditorReferenceType1Choice cdOrPrtry = objectFactory.createCreditorReferenceType1Choice();
+					tp.setCdOrPrtry(cdOrPrtry);
+					cdOrPrtry.setPrtry("QRR");
+
+					cdtrRefInf.setRef(QRReference);
 				}
 				else
 				{
-					final String validReference = StringUtils.trunc(replaceForbiddenChars(reference), 140, TruncateAt.STRING_START);
-					rmtInf.setUstrd(validReference);
+					// note: we use the structuredRemittanceInfo in ustrd, if we do SEPA (zahlart 5),
+					// because it's much less complicated
+					final String reference = Check.isEmpty(line.getStructuredRemittanceInfo(), true)
+							? line.getDescription()
+							: line.getStructuredRemittanceInfo();
```

Comment:
this could be written as
final String reference = StringUtils.trimBlankToNull(line.getStructuredRemittanceInfo()).orElseGet(line::getDescription);